File: modules/bar/calendar.ts
import { Widget, Gtk, PopupWindow, Utils } from "../../imports";
import options from "options"

const { Box, Label } = Widget;
const { execAsync } = Utils;
//const Calendar = Widget.subclass(Gtk.Calendar)
const { bar, datewin } = options;
const pos = datewin.position.bind();
const layout = Utils.derive([bar.position, datewin.position], (bar, qs) => 
    `${bar}-${qs}` as const,
);

const CalWidWin = () =>  PopupWindow({
    name: "calendar",
    className: "calpopwin",
    anchor: pos,
    transition: pos.as(pos => pos === "top" ? "slide_down" : "slide_up"),
    layer: "top",
    exclusivity: 'normal',
    keymode: 'on-demand',
    margins: [0,550],
    child: Box({
        className: "calendarbox",
        child: Widget.Calendar({
            showDayNames: true,
            showDetails: true,
            showHeading: true,
            showWeekNumbers: true,
            hpack: "center",
            vpack: "center",
        })
    })
});

export function Calendar() {
    App.addWindow(CalWidWin())
    layout.connect("changed", () => {
        App.removeWindow("calendar")
        App.addWindow(CalWidWin())
    })
}

File: Rust/Modules/Windows/Calendar.rs
use imports::{Widget, PopupWindow, Utils, Gtk, App, GLib};
use options;
use crate::Widgets::GridCalendar;

fn Calendar() {
    let bar = options::bar.position();
    let datewin = options::datewin.position();
    let pos = datewin;
    let layout = Utils::derive(
        [bar, datewin],
        |bar, qs| format!("{}-{}", bar, qs),
    );

    let CalWidWin = || {
        PopupWindow {
            name: "calendar",
            className: "calpopwin",
            anchor: ["top"],
            transition: pos.map(|pos| if pos == "top" { "slide_down" } else { "slide_up" }),
            layer: "top",
            exclusivity: "normal",
            keymode: "on-demand",
            child: Box {
                className: "calendarbox",
                child: GridCalendar(),
            },
        }
    };

    App::addWindow(CalWidWin());
    layout.connect("changed", || {
        App::removeWindow("calendar");
        App::addWindow(CalWidWin());
    });
}